apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/rpc/impl.ex

```diff
@@ -0,0 +1,172 @@
+defmodule Astarte.DataUpdaterPlant.RPC.Impl do
+  alias ElixirSense.Log
+  alias Astarte.DataUpdaterPlant.RPC.State
+  alias Astarte.DataUpdaterPlant.RPC.Device
+  alias Astarte.DataUpdaterPlant.RPC.Queries
+  require Logger
+
+  # State initialization by retrieving state via DB queries
+  def init_state() do
+    Logger.info("Initializing dup rpc handler state ...")
+    Logger.info("Fetching realms ...")
+
+    realms = retrieve_realms()
+    Logger.info("Retrieved realms: #{inspect(realms)}")
+
+    devices =
+      realms
+      |> Enum.reduce(%{}, fn realm, acc ->
+        Logger.info("Fetching connected devices for realm: #{realm} ...")
+
+        case retrieve_connected_devices(realm) do
+          {:ok, realm_devices} ->
+            Logger.info("Connected devices for realm #{realm} are: #{inspect(realm_devices)}")
+
+            Enum.reduce(realm_devices, acc, fn device, acc_map ->
+              Map.put(acc_map, device.device_id, device)
+            end)
+
+          {:error, reason} ->
+            Logger.error("Failed to fetch devices for realm #{realm}: #{inspect(reason)}")
+            acc
+        end
+      end)
+
+    %State{
+      devices: devices
+    }
+  end
+
+  def get_trigger_installation_scope(simple_trigger) do
+    case simple_trigger do
+      {:data_trigger, %Astarte.Core.Triggers.SimpleTriggersProtobuf.DataTrigger{} = data_trigger} ->
+        cond do
+          Map.has_key?(data_trigger, :group_name) and not is_nil(data_trigger.group_name) ->
+            {:data_trigger_group, data_trigger.group_name}
+
+          Map.has_key?(data_trigger, :device_id) and not is_nil(data_trigger.device_id) ->
+            {:data_trigger_device, data_trigger.device_id}
+
+          true ->
+            {:all, nil}
+        end
+
+      {:device_trigger,
+       %Astarte.Core.Triggers.SimpleTriggersProtobuf.DeviceTrigger{device_id: device_id}} ->
+        {:device, device_id}
+
+      _ ->
+        {:unknown, nil}
```

Comment:
If we don't expect other possible matches, we can just remove this pattern match and let the code crash right here (because no pattern matches) to indicate that we were not expecting other types of `simple_trigger`